compiler/lib/global_flow.ml

```diff
@@ -0,0 +1,624 @@
+  let variable_possibly_mutable = Array.make nv false in
+  let may_escape = Array.make nv No in
+  let possibly_mutable = Array.make nv false in
+  let st =
+    { vars
+    ; deps
+    ; defs
+    ; return_values = rets
+    ; variable_may_escape
+    ; variable_possibly_mutable
+    ; may_escape
+    ; possibly_mutable
+    ; known_cases = Hashtbl.create 16
+    ; applied_functions = Hashtbl.create 16
+    }
+  in
+  program_deps st p;
+  if times ()
+  then Format.eprintf "      global flow analysis (initialize): %a@." Timer.print t1;
+  let t2 = Timer.make () in
+  let approximation = solver st in
+  if times ()
+  then Format.eprintf "      global flow analysis (solve): %a@." Timer.print t2;
+  if times () then Format.eprintf "    global flow analysis: %a@." Timer.print t;
+  if debug ()
+  then
+    Var.ISet.iter
+      (fun x ->
+        let s = Var.Tbl.get approximation x in
+        if true || not (Poly.( = ) s Domain.bot)
+        then
+          Format.eprintf
+            "%a: %a@."
+            Var.print
+            x
+            (fun f a ->
+              match a with
+              | Top -> Format.fprintf f "top"
+              | Values { known; others } ->
+                  Format.fprintf
+                    f
+                    "{%a/%b} mut:%b vmut:%b esc:%s"
+                    (Format.pp_print_list
+                       ~pp_sep:(fun f () -> Format.fprintf f ", ")
+                       (fun f x ->
+                         Format.fprintf
+                           f
+                           "%a(%s)"
+                           Var.print
+                           x
+                           (match st.defs.(Var.idx x) with
+                           | Expr (Closure _) -> "C"
+                           | Expr (Block _) ->
+                               "B"
+                               ^
+                               if Poly.equal st.may_escape.(Var.idx x) Escape
+                               then "X"
+                               else ""
+                           | _ -> "O")))
+                    (Var.Set.elements known)
+                    others
+                    st.possibly_mutable.(Var.idx x)
+                    st.variable_possibly_mutable.(Var.idx x)
+                    (match st.may_escape.(Var.idx x) with
+                    | Escape -> "Y"
+                    | Escape_constant -> "y"
+                    | No -> "n"))
+            s)
+      vars;
+  { info_defs = defs
+  ; info_approximation = approximation
+  ; info_may_escape = Array.map ~f:(fun s -> Poly.(s <> No)) may_escape
+  }
+
+let exact_call info f n =
+  match Var.Tbl.get info.info_approximation f with
+  | Top | Values { others = true; _ } -> false
+  | Values { known; others = false } ->
+      Var.Set.for_all
+        (fun g ->
+          match info.info_defs.(Var.idx g) with
+          | Expr (Closure (params, _)) -> List.length params = n
+          | Expr (Block _) -> true
```

Comment:
Why is this true ?

Reply:
We know that `g` will never actually be a block. So, we can ignore this value.